runtime/build.rs

```diff
@@ -44,8 +40,28 @@ fn create_snapshot(
 }
 
 fn create_runtime_snapshot(snapshot_path: &Path, files: Vec<PathBuf>) {
+  let deno_modules: Vec<Box<dyn JsRuntimeModule>> = vec![
+    Box::new(deno_webidl::init()),
+    Box::new(deno_console::init()),
+    Box::new(deno_url::init()),
+    Box::new(deno_web::init()),
+    Box::new(deno_fetch::init::<deno_fetch::NoFetchPermissions>(
+      "".to_owned(),
+      None,
+    )),
+    Box::new(
+      deno_websocket::init::<deno_websocket::NoWebSocketPermissions>(
+        "".to_owned(),
+        None,
+      ),
+    ),
+    Box::new(deno_crypto::init(None)),
+    Box::new(deno_webgpu::init(false)),
+  ];
```

Comment:
I have some reservations about this - it requires to provide some state during snapshotting process which wasn't necessary before; it doesn't look so bad, but I can imagine situation with 20 crates and providing dummy values in such case might get complicated. Maybe there should be separate methods for obtaining JS sources and initializing the state?

Reply:
This is indeed a small quirk and could be avoided, since JS provided by extensions isn't dependant on input args (currently and probably shouldn't). Both ops and opstate (and middleware once added) depend on args.

Splitting JS from ops would break the conceptual "bundle" that extensions provide.

Do you have a better idea on how to do this without breaking/leaking the extension abstraction ?

Currently I think dummy-args are a lesser evil, not ideal but better than a leaky abstraction.

Part of the goal of extensions, is to streamline snapshotting, so 3rd parties can easily create their own snapshots with relative ease.

Reply:
@bartlomieju I think snapshotting JsRuntimes should be less of an edge case moving forward, in https://github.com/denoland/deno/pull/10308 I suggest allowing opcalls prior to snapshotting.

So rather than breaking the abstraction of runtime extensions for the sake of snapshots I think we should handle snapshots more gracefully and allow snapshotting fully initialized runtimes (with fully initialized extensions, js, ops, ...).

Reply:
Snapshotting should no longer be an edge case since the op-layer rewrite and https://github.com/denoland/deno/pull/10308